### ISSUE TITLE
What does it mean "Replaying full-use information from the cache"?

### ISSUE THREAD

**Issue by carlosgalvezp** [include-what-you-use/include-what-you-use#1291]
Hi!

We are tracking down an IWYU error (IWYU wrongly wants to remove `#include <iterator>` even when we are using `std::end` in the code. 

We raised the verbosity to highest level and tried to figure out what header does IWYU think that `std::end` belongs to. However we cannot find any of the typical messages `Marked full-info use ...`. The only thing we find for `std::end` is:

```
(Replaying full-use information from the cache for std::end)
```

Which doesn't quite tell us what header does IWYU think that `std::end` should come from.

Do you have any insight into why it might be doing that? Understanding this might help us create a minimal example that we can share (currently it's a fair amount of complex templated code).

Thanks!

**Comment by carlosgalvezp** [comment]
As it turns out I'm stumbling into the same error I reported here :) 
https://github.com/include-what-you-use/include-what-you-use/issues/1250

I'd be interested however to understand why there's to mention of `std::end` in the IWYU logs.

**Comment by kimgr** [comment]
In short, I think you're running into the unresolvedlookupexpr issue tag. The template is never instantiated and std::end is dependent ok the template arg, so iwyu never even sees it.

But I haven't had a chance to verify that yet, could be wrong.

**Comment by kimgr** [comment]
`Replaying full-use information from cache` is marked with the following comment in IWYU sources:
```
    // If we have cached the reporting done for this decl before,
    // report again (but with the new caller_loc this time).
    // Otherwise, for all reporting done in the rest of this scope,
    // store in the cache for this function.
```

So I think in this case `std::end` has already been evaluated in a way that would expand to the same thing, and the uses are replayed from a cache of the last time it was processed. I'm not 100% sure how that works, there's a bunch of moving parts there. Your example in #1293 does not appear to have more than one `std::end` call/instantiation, so not sure if you're seeing it with that reduced example too?

**Comment by carlosgalvezp** [comment]
Yeah I got that feeling as well, but then I got confused because I couldn't find "the first time" that `std::end` was processed before it entered into the cache. I'll try to dig deeper in the logs and share some more :)
